### PR TITLE
Deploy heterogeneous cluster across Kubernetes clusters and support to deploy heterogenous pump and ticdc

### LINKED ISSUE #3169: Deploy heterogeneous cluster across Kubernetes clusters
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by KanShiori**
### What is changed and how does it work?

Close #3169 #4246

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [x] E2E test 
- [x] Manual test 

* Deploy heterogeneous cluster across Kubernetes clusters
* Deploy heterogeneous cluster in one Kubernetes cluster
* Deploy heterogeneous cluster with TLS-enabled across Kubernetes clusters
* Deploy heterogeneous  pump and ticdc in one Kubernetes cluster

- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Fix the issue that heterogeneous cluster can't start when deploy it across kubernete clusters, and support to deploy heterogeneous  pump and ticdc
```

### CODE REVIEW

File: tests/e2e/tidbcluster/across-kubernetes.go [hunk 1]
```diff
@@ -326,6 +342,24 @@ var _ = ginkgo.Describe("[Across Kubernetes]", func() {
...
+			tcName1, tcName2 := "tls-cluster-1", "tls-cluster-2-without-pd"
+			tc1 := GetTCForAcrossKubernetes(ns1, tcName1, version, clusterDomain, nil)
+			tc2 := GetTCForAcrossKubernetes(ns2, tcName2, version, clusterDomain, tc1)
+			tc2.Spec.PD = nil
+
+			ginkgo.By("Prepare TLS resources for clusters")
+			MustPrepareXK8sTLSResources(genericCli, tc1, []*v1alpha1.TidbCluster{tc2})
+
+			ginkgo.By("Deploy all clusters and wait status to be ready")
+			MustCreateXK8sTCWithComponentsReady(genericCli, oa, []*v1alpha1.TidbCluster{tc1, tc2}, false)
```
Comment (csuzhangxc) [inline tests/e2e/tidbcluster/across-kubernetes.go:357]: why not `true` for `tlsEnabled?

File: pkg/apis/pingcap/v1alpha1/tidbcluster.go [hunk 2]
```diff
@@ -931,6 +931,18 @@ func (tc *TidbCluster) TiCDCLogLevel() string {
...
+func (tc *TidbCluster) Heterogeneous() bool {
+	return tc.Spec.Cluster != nil && len(tc.Spec.Cluster.Name) > 0
+}
+
+func (tc *TidbCluster) WithoutLocalPD() bool {
+	return tc.Spec.PD == nil
+}
+
+func (tc *TidbCluster) HeterogeneousWithRemote() bool {
+	return tc.Heterogeneous() && tc.Spec.ClusterDomain != ""
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/tidbcluster.go:943]: We should check `tc.spec.cluster.clusterDomain`?

Comment (KanShiori) [reply pkg/apis/pingcap/v1alpha1/tidbcluster.go:943]: yes... fix it later

File: tests/e2e/tidbcluster/across-kubernetes.go (hunk 1 above)
Comment (KanShiori) [reply tests/e2e/tidbcluster/across-kubernetes.go]: mistake... fix it

File: pkg/monitor/monitor/monitor_manager.go [hunk 3]
```diff
@@ -99,7 +99,7 @@ func (m *MonitorManager) SyncMonitor(monitor *v1alpha1.TidbMonitor) error {
 			}
 		}
 
-		if firstTc == nil && !tc.HeterogeneousWithoutLocalPD() {
+		if firstTc == nil && !(tc.Heterogeneous() && tc.WithoutLocalPD()) {
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/monitor_manager.go:102]:
```suggestion
		if firstTc == nil && !tc.WithoutLocalPD() {
```

File: tests/e2e/tidbcluster/across-kubernetes.go [hunk 4]
```diff
@@ -212,6 +212,22 @@ var _ = ginkgo.Describe("[Across Kubernetes]", func() {
 			framework.ExpectNoError(err, "tc1 is not connectable")
 		})
 
+		ginkgo.It("Deploy clusters and cluster-2 do not have PD", func() {
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/across-kubernetes.go:215]: Please also help cover all the components in the heterogeneous cases in tidbcluster.go.

File: pkg/discovery/discovery.go [hunk 5]
```diff
@@ -252,6 +254,16 @@ func (d *tidbDiscovery) VerifyPDEndpoint(pdURL string) (string, error) {
 		return pdURL, err
 	}
 
+	// if local pd doesn't exist, return target cluster pd peer addr
+	if tc.Heterogeneous() && tc.WithoutLocalPD() {
```
Comment (DanielZhangQD) [inline pkg/discovery/discovery.go:258]: If it's not remote heterogeneous cluster, no need to use the PD peer address, so we'd better check remote here.

File: pkg/manager/member/ticdc_member_manager.go [hunk 6]
```diff
@@ -358,10 +358,10 @@ func getNewTiCDCStatefulSet(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap) (*ap
 		cmdArgs = append(cmdArgs, fmt.Sprintf("--ca=%s", path.Join(ticdcCertPath, corev1.ServiceAccountRootCAKey)))
 		cmdArgs = append(cmdArgs, fmt.Sprintf("--cert=%s", path.Join(ticdcCertPath, corev1.TLSCertKey)))
 		cmdArgs = append(cmdArgs, fmt.Sprintf("--key=%s", path.Join(ticdcCertPath, corev1.TLSPrivateKeyKey)))
-		if tc.Spec.ClusterDomain == "" {
-			cmdArgs = append(cmdArgs, fmt.Sprintf("--pd=https://%s-pd:2379", tcName))
-		} else {
+		if tc.HeterogeneousWithRemote() {
 			cmdArgs = append(cmdArgs, "--pd=${result}")
+		} else {
+			cmdArgs = append(cmdArgs, fmt.Sprintf("--pd=https://%s-pd:2379", tcName))
```
Comment (DanielZhangQD) [inline pkg/manager/member/ticdc_member_manager.go:364]: This is incorrect for the local heterogeneous case.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 7]
```diff
@@ -1811,10 +1811,15 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
 			ginkgo.By("Deploy heterogeneous tc")
 			heterogeneousTc := fixture.GetTidbCluster(ns, "heterogeneous", utilimage.TiDBLatest)
 			heterogeneousTc = fixture.AddTiFlashForTidbCluster(heterogeneousTc)
+			heterogeneousTc = fixture.AddPumpForTidbCluster(heterogeneousTc)
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1814]: There is also a TLS case.

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by csuzhangxc — APPROVED** [review]

**Pull request merged**